[PVR]  Connection eventlog changes

emveepee opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->

In investigation why PR #22212 did not always show the friendly name a few functional changes are required

- if the pvr client has not provided name string don't show Unknown show the generic addon name
- use the pvr client provided event log name if supplied
- log pvr connection messages
- create name strings from pvr even when the state is not connected.

<!--- Describe your change in detail here. -->

Currently the CPVRManager::ConnectionStateChange(CPVRClient* client, const std::string& connectString, PVR_CONNECTION_STATE state, const std::string&)

1. PVR does not use the second paramater for the user supplied event title.

Change will use this value if suppled.

2. PVR Does not output connection string unless all notifications are logged.

Change PVR connection messages to Basic level.
Given the nature of this change I will submit a few PR's to address what looks like incorret of clear use of ConnectionStateChange() in
pvr.vdr.vnsi and pvr.nextpvr  pvr.filmon could be changed since the URL for the title might seem odd.

3. PVR Fails to use a useful title if ConnectionStateChange() is called before the addon is created.

a. Fallback to the generic name used before PR #22212 if the friendly name is still the default.
b. A second commit is provided to attempt to get name information from the addon in all states.  It is not clear why PVRClient::SetConnectionState() did not always call GetAddonProperties() so the function was split into name string and other functions.  GetAddonProperties() could be restructured and always called too.  GetAddonProperties() is actually an odd function as a failure in it will cause the addon to not load but only when SetConnectionState is called and passed PVR_CONNECTION_STATE_CONNECTED.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->

Logging information event messages requires all notifications to be logged.  The friendlyName is not always friendly or clear.  When connections are dropped there is an error logged but by default there is no corresponding log after reconnecting.

<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Tested with pvr.nextpvr with several state changes.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Better information on the event log.  Better state information on connections, especially after disconnects.


## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [x] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed## Description

ksooo commented on xbmc/pvr/addons/PVRClient.cpp:
```diff
@@ -308,58 +311,7 @@ bool CPVRClient::GetAddonProperties()
-
-  if (retVal != PVR_ERROR_NO_ERROR && retVal != PVR_ERROR_NOT_IMPLEMENTED)
-    return false;
+  const std::string strFriendlyName = m_strFriendlyName;
```
this is problematic as you are accessing the member without mutex protection.

emveepee replied:
Right.

ksooo commented on xbmc/pvr/addons/PVRClient.cpp:
```diff
@@ -447,15 +399,81 @@ bool CPVRClient::GetAddonProperties()
+  /* display name = backend name:connection string */
+  strFriendlyName = strlen(strConnectionString) == 0 // connection string is optional
+                        ? strBackendName
+                        : StringUtils::Format("{}:{}", strBackendName, strConnectionString);
```
As an idea how to solve the mutex protection problem (and maybe also the localization problem), get rid of the `m_strFriendlyName` member and extend `GetFriendlyName` to calculate the value on demand, from backend name and what not.

emveepee replied:
I will look into that change, that makes the most sense to me

emveepee replied:
Ok changed.  It still doesn't use a mutex but the other Get name functions don't either.

emveepee commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -847,6 +847,15 @@ void CPVRClients::ConnectionStateChange(CPVRClient* client,
+  std::string strTitle;
+  if (!strEventTitle.empty())
+    strTitle = strEventTitle;
+  else if (client->GetFriendlyName() != g_localizeStrings.Get(13205))
```
I didn't want to duplicate the define as there is no shared header between PVRClient and PVRClients so things could still go wrong so this keeps the two in sync at runtime

ksooo replied:
>  so this keeps the two in sync at runtime

 No, it does not. CPVRClient instance caches the string #1305 value. If it is done when locale is en_GB, as an example, the string reads "Unknown" (or whatever). Then, user changes the locale to de_DE in Kodi settings. g_localizeStrings.Get(13205) evakuates to "Unbekannt" then and the code above does not work as you intend.

emveepee replied:
Thanks I didn't understand that.

emveepee commented on xbmc/pvr/PVREventLogJob.h:
```diff
@@ -20,24 +21,56 @@ class CPVREventLogJob : public CJob
     Event(bool bNotifyUser, bool bError, const std::string& label, const std::string& msg, const std::string& icon)
-    : m_bNotifyUser(bNotifyUser), m_bError(bError), m_label(label), m_msg(msg), m_icon(icon) {}
+      : m_bNotifyUser(bNotifyUser),
+        m_eLevel(bError ? EventLevel::Error : EventLevel::Information),
```
I didn't change the other calls to PVREventLog that are still using bError.  I could change them to use the enum if you prefer.

ksooo replied:
Yes, please change the other calls. So we need only one ctor. Simplifies things a bit.

emveepee replied:
Done

emveepee commented on xbmc/pvr/PVREventLogJob.h:
```diff
@@ -9,6 +9,7 @@
 #pragma once
 
 #include "utils/Job.h"
+#include "events/EventLog.h"
```
Yes I applied the global diff and instead need to apply it to the first commit.

ksooo replied:
Just squash the two commits down to one. The changes are not too complicated and can be reviewed even if they are all in one commit.

emveepee replied:
Done.  I did add a couple of more Name() to GetFriendlyName() conversions that made sense.  I don't bother on the scannning or when the addon fails.

Commit:
PVR State change event logging

PR #22212 fix for early call,  create eventlog of any type, log connection as a basic event log call.  Use  addon provide event log title.  Dynamic GetFriendlyName()  replace a few more generic Name() calls.

ksooo commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -847,17 +847,23 @@ void CPVRClients::ConnectionStateChange(CPVRClient* client,
     strMsg = strMessage;
   else
-    strMsg = g_localizeStrings.Get(iMsg);
+    strMsg = client->GetFriendlyName();
```
This change looks definitely wrong. `iMsg` contains the id of the various connection status strings, like "connecting", "connection lost" etc...

emveepee replied:
It is definitely wrong and corrected but I still need to retest the state changes.

ksooo commented on xbmc/pvr/addons/PVRClient.cpp:
```diff
@@ -482,7 +493,13 @@ const std::string& CPVRClient::GetConnectionString() const
+  /* display name = backend name:connection string */
+  const std::string strFriendlyName =
+      GetConnectionString().empty() // connection string is optional
+          ? GetBackendName() != g_localizeStrings.Get(13205) ? GetBackendName() : Name()
```
This does not work. String 13025 is "Unknown" (capital U) and `DEFAULT_INFO_STRING_VALUE` is "unknown" (small u"). And in other languages the two strings may differ even more. For de_de 13205 is "Unbekannt", for example. Even though both strings would be accidentally equal, this code would break as soon as somebody either changes the value of `DEFAULT_INFO_STRING_VALUE` or string 13205. 

What you probably want to do is `GetBackendName() != DEFAULT_INFO_STRING_VALUE`?

emveepee replied:
Yes is what I wanted to do, though about it while hiking in the snow and couldn't change it.

ksooo commented:
@emveepee  Maybe https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/guilib/PVRGUIActionsClients.cpp#L76 should also be changed to use `GetFriendlyName` ?

ksooo commented:
Regarading `GetFriendlyName`semantics and where the string is used : Would it make more sense to change implementation to return the instance name or for single instance the add-on name?

ksooo commented on xbmc/pvr/addons/PVRClient.cpp:
```diff
@@ -447,15 +398,75 @@ bool CPVRClient::GetAddonProperties()
+  char strConnectionString[PVR_ADDON_NAME_STRING_LENGTH] = {};
+  char strBackendVersion[PVR_ADDON_NAME_STRING_LENGTH] = {};
+  char strBackendHostname[PVR_ADDON_NAME_STRING_LENGTH] = {};
+  std::string strFriendlyName;
```
Not used, can be removed.

emveepee replied:
Right 415 removed.

ksooo commented on xbmc/pvr/addons/PVRClient.cpp:
```diff
@@ -447,15 +398,75 @@ bool CPVRClient::GetAddonProperties()
   m_strBackendHostname = strBackendHostname;
-  m_timertypes = timerTypes;
 
   return retVal == PVR_ERROR_NO_ERROR;
```
This should read `return true;`. Some of the API methods (e.g. GetBackendHostname) are optional and can return `PVR_ERROR_NOT_IMPLEMENTED`. If  that happens `retVal` here will contain `PVR_ERROR_NOT_IMPLEMENTED` which is not an error, but you would return false and the add-on will not work. Happens for example for pvr.plutotv.

ksooo commented on xbmc/pvr/addons/PVRClient.cpp:
```diff
@@ -447,15 +398,75 @@ bool CPVRClient::GetAddonProperties()
   m_strBackendHostname = strBackendHostname;
-  m_timertypes = timerTypes;
 
   return retVal == PVR_ERROR_NO_ERROR;
```
This should read return true;. Some of the API methods (e.g. GetBackendHostname) are optional and can return PVR_ERROR_NOT_IMPLEMENTED. If that happens retVal here will contain PVR_ERROR_NOT_IMPLEMENTED which is not an error, but you would return false and the add-on will not work. Happens for example for pvr.plutotv.

emveepee replied:
Right you are.  The AddonProperties logic is so odd if false is every returned the addon is disabled and cannot be automatically enabled.

emveepee commented:
> @emveepee Maybe https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/guilib/PVRGUIActionsClients.cpp#L76 should also be changed to use `GetFriendlyName` ?

I find that dialog is already quite busy 
![image](https://user-images.githubusercontent.com/2148031/209819746-f94a01fb-2793-43b6-8d08-80c7c4966046.png)

and the change makes it more complicated.  pvr.nextpvr followed the example of some others to include the state in the connection string and i guess it helps show which one might need the WOL but overall it is pretty ugly.

![image](https://user-images.githubusercontent.com/2148031/209819798-72fa13b8-b2f3-4f39-9eee-46f2ff1e4bb4.png)

emveepee commented:
> Regarding `GetFriendlyName`semantics and where the string is used : Would it make more sense to change implementation to return the instance name or for single instance the add-on name?
> 
> Something like this?
> 
> ```c++
> @@ -493,13 +492,15 @@ const std::string& CPVRClient::GetConnectionString() const
>  
>  const std::string CPVRClient::GetFriendlyName() const
>  {
> -  /* display name = backend name:connection string */
> -  const std::string strFriendlyName =
> -      GetConnectionString().empty() // connection string is optional
> -          ? GetBackendName() != DEFAULT_INFO_STRING_VALUE ? GetBackendName() : Name()
> -          : StringUtils::Format("{}:{}", GetBackendName(), GetConnectionString());
> +  if (Addon()->SupportsInstanceSettings())
> +  {
> +    std::string instanceName;
> +    Addon()->GetSettingString(ADDON_SETTING_INSTANCE_NAME_VALUE, instanceName, InstanceId());
> +    if (!instanceName.empty())
> +      return StringUtils::Format("{} ({})", Name(), instanceName);
> +  }
>  
> -  return strFriendlyName;
> +  return Name();
>  }
>  
>  PVR_ERROR CPVRClient::GetDriveSpace(uint64_t& iTotal, uint64_t& iUsed)
> ```

I do like the idea, except I prefer the BaseName over Name + InstanceName.  In my screen grab above that would replace NextPVR: with NextPVR PVR Client:ODROID N2 making it unnecessarily busy.  

Getting rid of the connection string definitely makes it more friendly so it could be used.

ksooo commented:
> I prefer the BaseName

What is BaseName?

emveepee commented:
> What is BaseName?

I meant BackendName.  What about this

```
const std::string CPVRClient::GetFriendlyName() const
{
  if (Addon()->SupportsInstanceSettings() 
        && GetBackendName() != DEFAULT_INFO_STRING_VALUE && !GetBackendName().empty())
  {
    return GetBackendName();
  }
  return Name();
}

emveepee commented:
I did try what you suggested and it is a bit wordier, but there does seem to be enough space now that connection junk is removed  Since the short form can be provide on the connection logs, I am happy with your suggestion.

![image](https://user-images.githubusercontent.com/2148031/209850359-196f53fe-8308-4b0a-9f9b-d27f39490a30.png)

I do see that pvr.hts overrides ConnectionStateChanged() with a connection string 

![image](https://user-images.githubusercontent.com/2148031/209850567-97238fb4-4f69-4862-975d-04172727b18c.png)

ksooo commented:
@emveepee what do you think, should we rename `GetFriendlyName` to `GetInstanceName`to reflect what it actually provides?

emveepee commented:
> @emveepee what do you think, should we rename `GetFriendlyName` to `GetInstanceName`to reflect what it actually provides?

Sure that is fine, or GetClientName() The reason I add the instance name to the backend name is it looks better on Settings->System information->PVR service.   That probably needs to bet changed too.

I also see pvr.hts doesn't add the instance name to the timer rule, pvr.nextpvr does.  Should that be done by core?  I don't think there is room for Name() in that dialog.

![image](https://user-images.githubusercontent.com/2148031/209860036-545b4c2e-d5c4-4c22-a892-cc2f179816eb.png)

ksooo commented:
> GetClientName() 

Also fine for me.

emveepee commented:
> > The reason I add the instance name to the backend name
> 
> That feels wrong as the backend name is unrelated to the instance name, strictly speaking.

Right the text needs to change from PVR Backend to PVR Client.
> 
> > add the instance name to the timer rule
> 
> Same. Mixing unrelated things into one string just because it looks "good" in the GUI is not a good idea. What, if I change the dialog to add the instance name ? ;-)

I would change the addon.   

> And I do not get why the instance name is of use in the timer rules selection dialog. Can you explain this, please? Does it show rules from different clients in one list? Don't remember this out of my head.

The rule including channel list is instance specific.

ksooo commented:
> The rule including channel list is instance specific.

Not sure this a bug in the timer rule selection dialog implementation of the timer settings dialog. A timer is related to a certain channel (or any channel) provided by a certain client. The channel list of the timer settings dialog is implemented correct and shows only the channels for one client. So, shouldn't the timer rules list also only contain rules from the client the channel belongs to? I think so.

emveepee commented:
> Not sure there is a bug in the timer type selection dialog implementation of the timer settings dialog. A timer type is related to a certain client. So, shouldn't the timer types list only contain types from the related client? I think so. The channel selection list is implemented correctly, as far as I can see.

To be clear I am talking about the two manual "Add timer..." options.  After you select Type you get a list of Types and these are mixed for all clients and client instances

Existing timers and rules are linked appropriately.

ksooo commented:
@emveepee As this PR is about event log changes I would like to postpone the timer settings matters (and other things related to denote the client instances properly) to another PR.

What I would like to change is the content of the connection state change event log entries. I would always use the friendly name as event title and append the add-on supplied message to the state change string, like this:

```c++
diff --git a/xbmc/pvr/addons/PVRClients.cpp b/xbmc/pvr/addons/PVRClients.cpp
index f666ac5530..2be7ba8868 100644
--- a/xbmc/pvr/addons/PVRClients.cpp
+++ b/xbmc/pvr/addons/PVRClients.cpp
@@ -23,6 +23,7 @@
 #include "pvr/channels/PVRChannelGroupInternal.h"
 #include "pvr/guilib/PVRGUIProgressHandler.h"
 #include "utils/JobManager.h"
+#include "utils/stringUtils.h"
 #include "utils/log.h"
 
 #include <algorithm>
@@ -808,7 +809,7 @@ void CPVRClients::OnPowerSavingDeactivated()
 }
 
 void CPVRClients::ConnectionStateChange(CPVRClient* client,
-                                        const std::string& strEventTitle,
+                                        const std::string& strConnectionString,
                                         PVR_CONNECTION_STATE newState,
                                         const std::string& strMessage)
 {
@@ -860,13 +861,6 @@ void CPVRClients::ConnectionStateChange(CPVRClient* client,
       return;
   }
 
-  // Use addon-supplied eventlog title if present
-  std::string strTitle;
-  if (!strEventTitle.empty())
-    strTitle = strEventTitle;
-  else
-    strTitle = client->GetFriendlyName();
-
   // Use addon-supplied message, if present
   std::string strMsg;
   if (!strMessage.empty())
@@ -874,9 +868,13 @@ void CPVRClients::ConnectionStateChange(CPVRClient* client,
   else
     strMsg = g_localizeStrings.Get(iMsg);
 
+  if (!strConnectionString.empty())
+    strMsg = StringUtils::Format("{} ({})", strMsg, strConnectionString);
+
   // Notify user.
   CServiceBroker::GetJobManager()->AddJob(
-      new CPVREventLogJob(bNotify, eLevel, strTitle, strMsg, client->Icon()), nullptr);
+      new CPVREventLogJob(
+          bNotify, eLevel, client->GetFriendlyName(), strMsg, client->Icon()), nullptr);
 }
 
 namespace
```

emveepee commented:
@ksooo that works cleanly, I modified the .h too.   I won't have to modify the two PVR's either, since it is an improvement their too.

ksooo commented:
> #include "utils/stringUtils.h"

@emveepee build breaker due to typo : StringUtils.h, with capital S.

ksooo commented:
```diff
diff --git a/xbmc/pvr/addons/PVRClients.cpp b/xbmc/pvr/addons/PVRClients.cpp
index 2b2e0e12f2..78aafe1659 100644
--- a/xbmc/pvr/addons/PVRClients.cpp
+++ b/xbmc/pvr/addons/PVRClients.cpp
@@ -23,8 +23,8 @@
 #include "pvr/channels/PVRChannelGroupInternal.h"
 #include "pvr/guilib/PVRGUIProgressHandler.h"
 #include "utils/JobManager.h"
+#include "utils/StringUtils.h"
 #include "utils/log.h"
-#include "utils/stringUtils.h"
 
 #include <algorithm>
 #include <functional>
```

emveepee commented:
> @emveepee build breaker due to typo : StringUtils.h, with capital S.

Sorry, it was compiling on Windows and I didn't notice.

ksooo commented on xbmc/pvr/addons/PVRClient.cpp:
```diff
@@ -243,6 +242,11 @@ void CPVRClient::SetConnectionState(PVR_CONNECTION_STATE state)
+  else
+  {
+    if (!GetAddonNameStringProperties())
+      CLog::LogF(LOGINFO, "Cannot read PVR client name string properties");
```
This should be a LOGERROR, right?

emveepee replied:
I can actually see no use case for setting these strings to return an error.  These wouldn't have been logged before in these state so the info is to the developer.

ksooo replied:
> I can actually see no use case for setting these strings to return an error.

Those strings are obtained via add-on API calls, which can return any error. And in this case `GetAddonNameStringProperties` returns false which should be logged as `LOGERROR` here. `LOGINFO`is for information, but not to make somebody aware of an error.

emveepee replied:
I made the change but I don't consider GetBackendVersion() returning PVR_ERROR_NOT_IMPLEMENTED when it is not connected is an error.  Even when connected it is questionable but I didn't want to change something that causes an add-on about right now.

ksooo commented on xbmc/pvr/PVREventLogJob.h:
```diff
@@ -19,25 +20,41 @@ class CPVREventLogJob : public CJob
   {
     bool m_bNotifyUser;
-    bool m_bError;
+    EventLevel m_eLevel = EventLevel::Information;
```
`EventLevel m_eLevel{EventLevel::Information};`

ksooo commented on xbmc/pvr/addons/PVRClient.cpp:
```diff
@@ -308,58 +310,7 @@ bool CPVRClient::GetAddonProperties()
-
-  if (retVal != PVR_ERROR_NO_ERROR && retVal != PVR_ERROR_NOT_IMPLEMENTED)
-    return false;
+  const std::string strFriendlyName = GetFriendlyName();
```
This variable can be avoided:

```diff
diff --git a/xbmc/pvr/addons/PVRClient.cpp b/xbmc/pvr/addons/PVRClient.cpp
index b44da852bd..d5c4947149 100644
--- a/xbmc/pvr/addons/PVRClient.cpp
+++ b/xbmc/pvr/addons/PVRClient.cpp
@@ -310,12 +310,11 @@ bool CPVRClient::GetAddonProperties()
 
   if (retVal != PVR_ERROR_NO_ERROR)
     return false;
-  const std::string strFriendlyName = GetFriendlyName();
 
   /* timer types */
   retVal = DoAddonCall(
       __func__,
-      [this, strFriendlyName, &addonCapabilities, &timerTypes](const AddonInstance* addon) {
+      [this, &addonCapabilities, &timerTypes](const AddonInstance* addon) {
         std::unique_ptr<PVR_TIMER_TYPE[]> types_array(
             new PVR_TIMER_TYPE[PVR_ADDON_TIMERTYPE_ARRAY_SIZE]);
         int size = PVR_ADDON_TIMERTYPE_ARRAY_SIZE;
@@ -328,7 +327,7 @@ bool CPVRClient::GetAddonProperties()
           CLog::LogF(LOGWARNING,
                      "Add-on {} does not support timer types. It will work, but not benefit from "
                      "the timer features introduced with PVR Addon API 2.0.0",
-                     strFriendlyName);
+                     GetFriendlyName());
 
           // Create standard timer types (mostly) matching the timer functionality available in Isengard.
           // This is for migration only and does not make changes to the addons obsolete. Addons should
```

ksooo commented on xbmc/pvr/addons/PVRClient.h:
```diff
@@ -165,7 +165,7 @@ class CPVRClient : public ADDON::IAddonInstanceHandler
   /*!
    * @return A friendly name for this add-on that can be used in log messages.
```
Mind changing the comment to reflect that the friendly name is also usable for GUI purposes?

emveepee replied:
Sure that reminds me the sometimes ID() is used for log messages and it doesn't show the instance but that is another PR

ksooo reviewed (approved):
Thanks.

ksooo commented:
@emveepee mind opening a backport PR?

ksooo commented:
@emveepee ping. If we want this in Nexus, we need the backport PR now. Please open it.

emveepee commented:
> @emveepee ping. If we want this in Nexus, we need the backport PR now. Please open it.

What about https://github.com/xbmc/xbmc/pull/22366

ksooo commented:
Ooops, missed it. Sorry.
